test(docs-infra): several small improvements in AIO & docs examples tests

gkalpak opened the pull request:
This PR contains several small improvements for the AIO app and the docs examples tests (mostly related to CI).
See the individual commits for details.

mary-poppins commented:
You can preview 6c4cd94 at https://pr43683-6c4cd94.ngbuilds.io/.

mary-poppins commented:
You can preview 8a2d38e at https://pr43683-8a2d38e.ngbuilds.io/.

mary-poppins commented:
You can preview 3d1355a at https://pr43683-3d1355a.ngbuilds.io/.

mary-poppins commented:
You can preview f4ea05d at https://pr43683-f4ea05d.ngbuilds.io/.

crisbeto commented on .circleci/config.yml:
```diff
@@ -466,6 +476,8 @@ jobs:
       resource_class: xlarge
     parallelism: 5
+    environment:
+      RXJS_VERSION_7: '7.3.1'
```
Is there a chance for the version to resolve to something like `7.3.2`, breaking this assertion?

petebacondarwin replied:
AFAICT the `yarn add` uses this explicit version so this is what should be installed:

```
yarn --cwd aio/tools/examples/shared add rxjs@$RXJS_VERSION_7
```

crisbeto commented on aio/tools/examples/run-example-e2e.mjs:
```diff
@@ -346,18 +347,32 @@ function reportStatus(status, outputFile) {
+    };
+    const reject = err => {
+      clearTimeout(noOutputTimeoutId);
+      rejectFn(err);
```
Should we also run `proc.kill()` if it times out?

gkalpak replied:
I don't think it matters, because the process will exit as soon as the parent process exits. But it doesn't hurt to be more explicit, so I'll make sure the process is terminated when there's a timeout :+1: 

[1]: https://github.com/angular/angular/blob/f4ea05d3d216a3d80cb4679d534f0edfca669869/aio/tools/examples/run-example-e2e.mjs#L245

petebacondarwin commented on aio/tools/examples/run-example-e2e.mjs:
```diff
@@ -347,17 +347,15 @@ function reportStatus(status, outputFile) {
     } catch (e) {
       console.log(e);
       reject(e);
-      return {proc: null, promise};
```
Is this a safe change to make?
If the `proc = ...` line throws an error, then `proc` will not be defined; and if we do not return here we will try to access `proc.stdout` below.

gkalpak replied:
Ah, you're right. The return statement is needed (to stop executing the function body). It's just that the return value is ignored. I'll put back the return statement.

petebacondarwin commented on aio/tools/examples/run-example-e2e.mjs:
```diff
@@ -25,6 +25,7 @@ const CLI_SPEC_FILENAME = 'e2e/src/app.e2e-spec.ts';
 const EXAMPLE_CONFIG_FILENAME = 'example-config.json';
 const DEFAULT_CLI_EXAMPLE_PORT = 4200;
 const DEFAULT_CLI_SPECS_CONCURRENCY = 1;
+const MAX_NO_OUTPUT_TIMEOUT = 1000 * 60 * 5;  // 5 minutes
```
Typo in the commit message for this change:

> ... improve debugging when docs example tests fail without exi~~s~~ting

gkalpak replied:
😱

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -478,7 +490,15 @@ jobs:
       - run: yarn --cwd aio example-e2e --setup --local --cliSpecsConcurrency=5 --shard=${CIRCLE_NODE_INDEX}/${CIRCLE_NODE_TOTAL}
         # Run tests with RxJS v7.
         # (Exclude some examples that are not yet compatible with v7.)
```
I think this comment should be below the new content here (i.e. line 502)

gkalpak replied:
I think of these comments as applying to the whole section below (all three steps that are needed to test against RxJS v7). But I see how this can be confusing. I'll move the comments.

gkalpak replied:
Oh, I think you might be referring to just the last line (about excluding certain examples).
I'll just move that one (line 492), but lmk if you think it would be better to move both.

petebacondarwin replied:
Yes - exactly.

mary-poppins commented:
You can preview 7eb7a15 at https://pr43683-7eb7a15.ngbuilds.io/.

petebacondarwin reviewed (approved):
Reviewed-for: docs-infra size-tracking

crisbeto reviewed (approved):
LGTM

gkalpak committed:
refactor(docs-infra): remove redundant code from `run-example-e2e.mjs`

This commit removes some code that was unused or had no effect from the
`run-example-e2e.mjs` script.

gkalpak committed:
test(docs-infra): improve debugging when docs example tests fail without exiting

When the docs examples tests run with `--cliSpecsConcurrency` greater
than 1 (as happens on CI), the output of each process is [buffered][1]
(to avoid interleaved output from multiple, parallel processes) and is
only printed out once the process exits (either successfully or with an
error). However, in cases where a process did not exit, the buffered
output would be never printed out, thus making debugging the failure
harder. This is the case, for example, if a build error happens during
`ng e2e`. This can be seen in action in [this CI job][2], where the job
fails due to no output, but the error (which is an incompatible TS
version) is never printed out.

To make debugging such situations easier in the future, this commit
updates the `spawnExt()` helper to reject (causing the buffered output
to be printed out) if there is no output from the spawned process for
more than 5 minutes.

[1]: https://github.com/angular/angular/blob/c721135e370b34c840756bcfb22c8119b4c8c452/aio/tools/examples/run-example-e2e.mjs#L293
[2]: https://circleci.com/gh/angular/angular/1061751

gkalpak committed:
fixup! refactor(docs-infra): remove redundant code from `run-example-e2e.mjs`

gkalpak committed:
ci: use the local Angular packages when testing AIO/docs examples with RxJS v7

The `test_aio_local` and `test_docs_examples` CI jobs, test the AIO app
and the docs examples using the locally built Angular packages. They
also include steps that install RxJS v7 and test the examples against
that as well, to avoid regressions with RxJS v7.

However, it turns out that the `yarn add` command used to install
RxJS v7 caused yarn to overwrite the local Agular packages in
`node_modules/` with the ones from NPM.

This commit ensures that the local Angular packages are restored in
`node_modules/` after adding RxJS v7. It also adds a check to ensure
that the restoration of local Angular packages did not affect the
version of RxJS.

gkalpak committed:
ci: test AIO/docs examples against the latest RxJS 7.x version (v7.3.1)

Use the latest RxJS 7.x version (v7.3.1) with AIO and docs examples
tests to guard against regressions with RxJS v7.

gkalpak committed:
fixup! test(docs-infra): improve debugging when docs example tests fail without exiting

gkalpak committed:
fixup! ci: use the local Angular packages when testing AIO/docs examples with RxJS v7

mary-poppins commented:
You can preview f248a91 at https://pr43683-f248a91.ngbuilds.io/.

dylhunn commented:
This PR was merged into the repository by commit 23ac3f7cef6c7330bba128411ddfa9f5d08600c7.
